# Проект SQL

<div style=" background:#bdc3c7; padding: 20px"> <h7 style="color:black; margin-bottom:20px"> Артём, привет!)
 Мои комментарии и замечания далее по тексту помечены различными цветами:
    
---    
<div class="alert alert-block alert-danger">
Критическое замечание, которые следует исправить.
</div> 
    
    
---
    
<div class="alert alert-block alert-info">
Замечание\совет на будущее.
</div>
    
---
    
<div class="alert alert-block alert-success">
Когда всё сделано правильно.
</div>
    
---
    
p.s.: не удаляй мои замечания, если предстоит что то доработать в проекте.</h7>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Итог ревью - ты проделал большую работу. Запросы понятны и логичны, но не хватает совсем чуть чуть. Есть задания, которые тебе предстоит доработать, чтобы я смог принять твой проект. Если возникнут вопросы по моим комментариям, пиши здесь - обязательно отвечу. Буду ждать!
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Михаил, добрый день! Попробовал исправить запросы -- старые тоже оставил на месте. Исправленые запросы обозначил постфиксом _v1</div>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Привет) Все замечания учтены. Успехов в дальнейшем обучении!
</div>

<b>artyom-ivanov-spb@yandex.ru</b>

## Цель

Проанализировать базу данных

## Задачи 

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Исследование данных

- Исследуйте таблицы — выведите первые строки;

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
books = pd.io.sql.read_sql('books', con = engine) 
authors = pd.io.sql.read_sql('authors', con = engine) 
publishers = pd.io.sql.read_sql('publishers', con = engine) 
ratings = pd.io.sql.read_sql('ratings', con = engine) 
reviews = pd.io.sql.read_sql('reviews', con = engine)

In [3]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-warning">
<b>Комментарии от ревьюера:</b>

Здесь всё хорошо. Ты корректно выгрузил данные и посмотрел их содержимое. Не помешало бы вывести еще информацию с помощью info(). Так у нас будет информация хотя бы о размерности датасетов
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Сейчас вывел ниже:</div>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Спасибо, так гораздо информативнее
</div>

In [8]:
bases = [books, authors, publishers, ratings, reviews]

for ind in bases:
    ind.info()
    print('=========================================')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null datetime64[ns]
publisher_id        1000 non-null int64
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-nu

- Посчитайте, сколько книг вышло после 1 января 2000 года;

Для дальнейшего удобства, создадим небольшую функцию, которая будет запускать скрипты запросов к удаленной базе.

In [9]:
def run_sql(sql_script):
    sql_query = pd.io.sql.read_sql(sql_script, con = engine)
    return sql_query

Посчитаем, сколько книг вышло после 1 января 2000 года.

In [10]:
query1 = "SELECT COUNT(title) from books where publication_date >= '2000-01-01'"

run_sql(query1)

count
0    821

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Всё верно. Это правильный ответ
</div>

<b>Итого</b> После 1 января 2000 года вышла 821 книга

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [11]:
query2 = 'SELECT title, COUNT(review_id), AVG(rating) FROM books LEFT JOIN reviews on books.book_id = reviews.book_id LEFT JOIN ratings on books.book_id = ratings.book_id GROUP BY title'

<div class="alert alert-block alert-warning">
<b>Комментарии от ревьюера:</b>

Советую использовать тройные кавычки, чтобы не запрос писать не в одну строку (так думаю будет удобнее, чем в одну строку). Пример ниже
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Благодарю, применил это оформление для всех запросов, которые исправил далее</div>

In [12]:
# код тимлида

query2 = '''  
            SELECT title, COUNT(review_id), AVG(rating)
            FROM books 
            LEFT JOIN reviews on books.book_id = reviews.book_id
            LEFT JOIN ratings on books.book_id = ratings.book_id
            GROUP BY title
          
         '''

In [13]:
run_sql(query2)

title  count       avg
0                            The Count of Monte Cristo    115  4.217391
1                              Count Zero (Sprawl  #2)      4  2.500000
2    The Botany of Desire: A Plant's-Eye View of th...      4  3.500000
3                                 The Poisonwood Bible    110  4.363636
4                                 The Canterbury Tales     18  3.333333
..                                                 ...    ...       ...
994                           Of Love and Other Demons      4  4.500000
995  In the Heart of the Sea: The Tragedy of the Wh...      9  3.333333
996               Welcome to Temptation (Dempseys  #1)      4  5.000000
997                      World's End (The Sandman  #8)      4  4.500000
998                                  Holes (Holes  #1)    155  3.967742

[999 rows x 3 columns]

<b>Итого</b> Нашли количество обзоров и среднюю оценку для каждой из книг books

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

<s>Увы, это неправильный вариант. Всего ревью в БД 2793, а у тебя в запросе сумма count 26167. Ты делаешь два джоина. Посмотри, что происходит с данными после каждого (без подсчета количества обозоров и среднего рейтинга) и тогда тебе станет понятно в чем ошибка
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Вероятно не учёл distinct, т.к. при соединении с рейтингами, рейтингов больше чем ревью. Т.е. средняя оценка высчиталась, получается, верно, но вот обзоры посчитались "не уникальные". Исправил запрос, вариант ниже.
</div>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Да, это верное решение. Теперь результат правильный)
</div>

In [14]:
query2_v1 = '''  

            SELECT title, COUNT(distinct review_id) as num_of_reviews, ROUND(AVG(rating),2) as average_rating
            FROM books 
            LEFT JOIN reviews on books.book_id = reviews.book_id
            LEFT JOIN ratings on books.book_id = ratings.book_id
            GROUP BY title

         '''

In [15]:
run_sql(query2_v1)

title  num_of_reviews  \
0                                         'Salem's Lot               2   
1                   1 000 Places to See Before You Die               1   
2    13 Little Blue Envelopes (Little Blue Envelope...               3   
3    1491: New Revelations of the Americas Before C...               2   
4                                                 1776               4   
..                                                 ...             ...   
994           Wyrd Sisters (Discworld  #6; Witches #2)               3   
995                        Xenocide (Ender's Saga  #3)               3   
996                                    Year of Wonders               4   
997                        You Suck (A Love Story  #2)               2   
998  Zen and the Art of Motorcycle Maintenance: An ...               4   

     average_rating  
0              3.67  
1              2.50  
2              4.67  
3              4.50  
4              4.00  
..              ...  
994            3.67  
995            3.40  
996            3.20  
997            4.50  
998            3.83  

[999 rows x 3 columns]

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [16]:
query3 = "SELECT publisher, count(title) FROM books LEFT JOIN publishers on books.publisher_id = publishers.publisher_id where books.num_pages > '50' GROUP BY publisher ORDER BY 2 desc"

In [17]:
run_sql(query3)

publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам
</div>

<b>Итого</b> Больше всего книг, где более 50 страниц, выпустило издательство Penguin Books

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [18]:
query4 = 'SELECT author, COUNT(rating_id), AVG(rating) from books LEFT JOIN authors on books.author_id = authors.author_id LEFT JOIN ratings on books.book_id = ratings.book_id GROUP BY author HAVING count(rating_id) > 50 ORDER BY 3 desc'

In [19]:
run_sql(query4)

author  count       avg
0                          J.K. Rowling/Mary GrandPré    312  4.288462
1                                     Agatha Christie     53  4.283019
2                   Markus Zusak/Cao Xuân Việt Khương     53  4.264151
3                                      J.R.R. Tolkien    166  4.240964
4                            Roald Dahl/Quentin Blake     62  4.209677
5                                   Louisa May Alcott     54  4.203704
6                                        Rick Riordan     84  4.130952
7                                       Arthur Golden     56  4.107143
8                                        Stephen King    106  4.009434
9                                        John Grisham     70  3.971429
10                                    William Golding     71  3.901408
11                                    Nicholas Sparks    111  3.882883
12                                       Jodi Picoult     76  3.881579
13                                    Sophie Kinsella     57  3.877193
14                                    James Patterson     64  3.859375
15                                      J.D. Salinger     98  3.846939
16           Paulo Coelho/Alan R. Clarke/Özdemir İnce     57  3.789474
17  William Shakespeare/Paul Werstine/Barbara A. M...     66  3.787879
18                                          Dan Brown    143  3.741259
19                                         Lois Lowry     65  3.738462
20            George Orwell/Boris Grabnar/Peter Škerl     74  3.729730
21                                    Stephenie Meyer    160  3.662500
22                                     John Steinbeck     73  3.643836

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Не совсем верный ответ. Во-первых, в условиях задачи нестрогое неравенство ">= 50", а у тебя строгое "> 50".
    
Во-вторых, ты подсчитавешь ```HAVING count(rating_id) > 50``` по авторам, а не по книгам (т.е. выводишь авторов у которых 50 и более оценок, а надо книги). Предлагаю первым подзапросом отфильтровать книги по условию задачи, а потом уже искать кто является их автором и рассчитывать показатели
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Попробовал сформировать запрос немного иначе, через подзапрос. Смущает немного обилие джоинов..
</div>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Результат корректный. Можно было подзапрос сделать from ratings и фильтровать book_id - это минус один join.
</div>

In [35]:
query4_v1 = '''

SELECT distinct author, ROUND(AVG(rating), 2) as average_rating
FROM books 
LEFT JOIN authors on books.author_id = authors.author_id
LEFT JOIN ratings on books.book_id = ratings.book_id
WHERE title IN (SELECT title
FROM books 
LEFT JOIN ratings on books.book_id = ratings.book_id 
GROUP BY title
HAVING count(rating_id) >= 50)
GROUP BY author
ORDER BY 2 DESC


'''

In [36]:
run_sql(query4_v1)

author  average_rating
0                          J.K. Rowling/Mary GrandPré            4.29
1                   Markus Zusak/Cao Xuân Việt Khương            4.26
2                                      J.R.R. Tolkien            4.25
3                                   Louisa May Alcott            4.19
4                                       Arthur Golden            4.11
5                                        Rick Riordan            4.08
6                                     William Golding            3.90
7                                       J.D. Salinger            3.83
8            Paulo Coelho/Alan R. Clarke/Özdemir İnce            3.79
9   William Shakespeare/Paul Werstine/Barbara A. M...            3.79
10                                         Lois Lowry            3.75
11                                          Dan Brown            3.74
12            George Orwell/Boris Grabnar/Peter Škerl            3.73
13                                    Stephenie Meyer            3.66
14                                     John Steinbeck            3.62

<b>Итого</b> Наилучшие оценки произведений у автора J.K. Rowling/Mary GrandPré

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [22]:
query5 = "SELECT ROUND(AVG(num_of_rws)) FROM (SELECT COUNT(reviews.review_id) as num_of_rws FROM reviews LEFT JOIN ratings ON reviews.book_id = ratings.book_id GROUP BY reviews.username HAVING COUNT(rating_id) >= 50) as cnt"

In [23]:
run_sql(query5)

round
0  166.0

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

<s>И здесь у тебя к сожалению ошибка. Всё гораздо проще. 
    
Я всегда рекомендую первым делом подзапросом выполнять условие задачи. В данном случае, тебе надо отфильтровать пользователей from ratings и оставить тех кто ставил больше 50 оценок. Так мы сразу отсекаем ненужные данные - это оптимальное решение. Т.е. на выходе подзапроса у нас просто имена пользователей username.

Далее нам надо посчитать сколько сделали они ревью. В reviews есть и username и review_id, просто потом фильтруем пользователей через ```where reviews.username in (*твой подзапрос*)```. Ну и найти общее среднее этих показателей. Попробуй)
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b>

Этот запрос сломал мне на жаре голову. Кажется понял, что имелось ввиду.. Здесь использовал строгий знак >, т.к. тут непосредственно "более 50".</div>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Согласен. Здесь надо быть внимательным к условиям задачи)
</div>

In [24]:
query5_v1 = '''

SELECT COUNT(review_id)/COUNT(distinct username) as avg_reviews_per_top_raters
FROM reviews 
WHERE username IN
(SELECT username
FROM ratings 
GROUP BY username
HAVING COUNT(rating_id) > 50)

'''

In [25]:
run_sql(query5_v1)

avg_reviews_per_top_raters
0                          24

<b>Итого</b> В среднем, пользователи, которые оставили более 50 оценок, напишут 166 обзоров.

<b>Вывод</b> 

1) после 1 января 2000 года вышла 821 книга

2) Для каждой книги нашли количество обзоров и среднюю оценку

3) Больше всего книг, где более 50 страниц, вышло в издательстве Penguin Books

4) Наилучшие оценки произведений, для которых есть более 50 оценок, у автора J.K. Rowling/Mary GrandPré

5) В среднем, пользователи, которые оставили более 50 оценок, напишут 24 обзора.